### PR TITLE
Allow squeezing of specific non-broadcastable axis

### LINKED ISSUE #705: Don't fail when squeezing specific non-broadcastable axis
### Description

The following snippet fails:

```python
import pytensor.tensor as pt

x = pt.vector("x")
pt.squeeze(x, axis=0)
# ValueError: Cannot drop a non-broadcastable dimension: [False], []
```

The user provided us information that this dimension will be of length 1, so we could add a `specify_broadcastable` and only then call `squeeze`. I think this behavior is more intuitive, and I don't see any drawbacks.

This can only be done when the user provides specific axis, not when axis=None

```python
pt.squeeze(pt.specify_broadcastable(x, 0), axis=0).eval({x: [0]}) 
# array(0.)
pt.squeeze(pt.specify_broadcastable(x, 0), axis=0).eval({x: [0, 1, 2]})
# AssertionError: SpecifyShape: dim 0 of input has shape 3, expected 1.
```

### PR DESCRIPTION
**Pull request by Dhruvanshu-Joshi**
## Description

Added support to allow squeezing of valid non-broadcastable axis. The axis will be checked and if it is valid and non_broadcastable, we first specify it as broadcastable and then squeeze it.
## Related Issue

- [x] Closes #705 
- [ ] Related to #

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

File: pytensor/tensor/extra_ops.py [hunk 1]
```diff
@@ -592,6 +593,10 @@ def squeeze(x, axis=None):
         # Nothing to do
         return _x
 
+    for i in axis:
```
Comment (ricardoV94) [inline pytensor/tensor/extra_ops.py:596]: Better to do a single call to specify_broadcastable with all the axis at once (smaller generated graph)

Also can you add a comment explaining what's going on for future devs reading the source?

File: tests/tensor/test_extra_ops.py [hunk 2]
```diff
@@ -463,13 +463,12 @@ def test_axis(self):
...
 
-    def test_invalid_axis(self):
-        # Test that trying to squeeze a non broadcastable dimension raises error
-        variable = TensorType(config.floatX, shape=(1, None))()
-        with pytest.raises(
-            ValueError, match="Cannot drop a non-broadcastable dimension"
-        ):
-            squeeze(variable, axis=1)
+    def test_non_broadcastable_axis(self):
+        # Test that we now allow squeezing of valid non-broadcastable dimension
```
Comment (ricardoV94) [inline tests/tensor/test_extra_ops.py:467]: Taking the now will not age :)

```suggestion
        # Test that we allow squeezing of valid non-broadcastable dimension
```

File: tests/tensor/test_extra_ops.py [hunk 3]
```diff
@@ -463,13 +463,12 @@ def test_axis(self):
...
+        x = pt.vector("x")
+        res = pt.squeeze(x, axis=0).eval({x: [0]})
+        expected_res = pt.squeeze(pt.specify_broadcastable(x, 0), axis=0).eval({x: [0]})
```
Comment (ricardoV94) [inline tests/tensor/test_extra_ops.py:470]: No need to compute expected_res like this, just assert below you get a 0

Comment (ricardoV94) [reply tests/tensor/test_extra_ops.py:470]: Can you also test when the input is invalid we get an expected error?

**Review by ricardoV94 — CHANGES_REQUESTED** [review]
Thanks, looks good, just some small suggestions

File: pytensor/tensor/extra_ops.py [hunk 4]
```diff
@@ -592,6 +593,13 @@ def squeeze(x, axis=None):
         # Nothing to do
         return _x
 
+    axes = ()
```
Comment (ricardoV94) [inline pytensor/tensor/extra_ops.py:596]: Confusing name, now we have `axes` and `axis`. We should call this intermediate variable something more obvious. Also all this can be done in a list comprehension, which is more pythonic/readbale:

Comment (ricardoV94) [reply pytensor/tensor/extra_ops.py:596]: Also still missing the comment that explains what is happening/why this is needed

File: pytensor/tensor/extra_ops.py [hunk 5]
```diff
@@ -593,12 +593,12 @@ def squeeze(x, axis=None):
         # Nothing to do
         return _x
 
-    axes = ()
-    for i in axis:
-        if i < _x.ndim and not _x.broadcastable[i]:
-            axes += (i,)
-    if axes:
-        _x = specify_broadcastable(_x, axes)
+    # if the axis is of length 1 but is not specifed as broadcastable, we first explicitly specify it as broadcastable
```
Comment (ricardoV94) [inline pytensor/tensor/extra_ops.py:596]:
```suggestion
    # `Dimshuffle` raises when we try to drop an axis that is not statically broadcastable.
    # We add a `specify_broadcastable` instead of raising.
```

File: pytensor/tensor/extra_ops.py [hunk 6]
```diff
@@ -593,12 +593,12 @@ def squeeze(x, axis=None):
...
+    non_broadcastable_axis = [
+        i for i in axis if i < _x.ndim and not _x.broadcastable[i]
+    ]
```
Comment (ricardoV94) [inline pytensor/tensor/extra_ops.py:599]: We can handle the `_x.ndim` case separately, and then you don't need that first check:

```suggestion
    if _x.ndim == 0:
        # Nothing could be squeezed
        return _x
        
    non_broadcastable_axis = [
        i for i in axis if not _x.broadcastable[i]
    ]
```

File: pytensor/tensor/shape.py [hunk 7]
```diff
@@ -994,6 +995,9 @@ def specify_broadcastable(x, *axes):
     if not axes:
         return x
 
+    if len(axes) == 1 and isinstance(axes[0], Collection):
+        axes = axes[0]
+
```
Comment (ricardoV94) [inline pytensor/tensor/shape.py:1000]: Why is this needed?
```suggestion
```

File: tests/tensor/test_extra_ops.py [hunk 8]
```diff
@@ -482,6 +479,14 @@ def test_scalar_input(self):
         ):
             squeeze(x, axis=1)
 
+    def test_invalid_input(self):
```
Comment (ricardoV94) [inline tests/tensor/test_extra_ops.py:482]: This can be merged with the new test above, it's the same graph and logic. You can compile the function once and test with the two inputs, valid and invalid

**Review by ricardoV94 — CHANGES_REQUESTED** [review]

File: pytensor/tensor/shape.py (hunk 7 above)
Comment (Dhruvanshu-Joshi) [reply pytensor/tensor/shape.py:1000]: the definition of specify_broadcastable is:
```
def specify_broadcastable(x, *axes):
```
The ```*``` converts axes into an iterable like (axes, ). Hence when we expect our non_broadcastable_axis list to be let's say [0], specify_broadcastable actually interprets it as ([0], ).

The comparison operation below hence throws an error:
https://github.com/pymc-devs/pytensor/blob/bcd81c7830d1ce0ccb2c16ac7010d23d1fcaafc3/pytensor/tensor/shape.py#L997

Comment (ricardoV94) [reply pytensor/tensor/shape.py:1000]: You can unpack the axis when you call specify_broadcastable woth *

```python
axes = [0, 1, 2]
pt.specify_broadcastable(x, *axes)
```

Works the same way regardless of whether there are multiple or a single axis

File: pytensor/tensor/extra_ops.py [hunk 9]
```diff
@@ -592,6 +593,19 @@ def squeeze(x, axis=None):
...
+        return _x
+
+    # `Dimshuffle` raises when we try to drop an axis that is not statically broadcastable.
+    # We add a `specify_broadcastable` instead of raising.
+    non_broadcastable_axis = [i for i in axis if not _x.broadcastable[i]]
+
+    if non_broadcastable_axis:
+        if not isinstance(non_broadcastable_axis, Collection):
+            non_broadcastable_axis = (non_broadcastable_axis,)
+        _x = specify_broadcastable(_x, *non_broadcastable_axis)
```
Comment (ricardoV94) [inline pytensor/tensor/extra_ops.py:607]: I think this suffices. The helper already has a check not to do anything if there are no axes
```suggestion
   _x = specify_broadcastable(_x, *non_broadcastable_axis)
```

**Review by ricardoV94 — APPROVED** [review]

File: tests/tensor/test_extra_ops.py [hunk 10]
```diff
@@ -482,6 +474,22 @@ def test_scalar_input(self):
...
+        axis = 0
+
+        f = pytensor.function([x], pt.squeeze(x, axis))
+
+        # Test that we allow squeezing of valid non-broadcastable dimension
+        assert f([0]) == 0
+
+        # Test that we cannot squeeze dimensions whose ength is greater than 1
+        with pytest.raises(
+            AssertionError,
```
Comment (ricardoV94) [inline tests/tensor/test_extra_ops.py:488]: The is failing because it doesn't raise an AssertionError but a ValueError?

Comment (Dhruvanshu-Joshi) [reply tests/tensor/test_extra_ops.py:488]: It does raise an assertion error. The regex pattern does not match strangely when using python 3.12 as evident [here](https://github.com/pymc-devs/pytensor/actions/runs/8750088177/job/24012954012?pr=713#step:6:4121) .

Comment (ricardoV94) [reply tests/tensor/test_extra_ops.py:488]: Is it version specific or do we just need to tweak the regex?

Comment (Dhruvanshu-Joshi) [reply tests/tensor/test_extra_ops.py:488]: It is version specific. Regex exp works fine with my local python v3.11.8

Comment (ricardoV94) [reply tests/tensor/test_extra_ops.py:488]: You may not be running locally with the same compile flag. Add this at the top of the test file and see if you can reproduce locally then:

```python
import pytensor
pytensor.config.mode = "FAST_COMPILE"
```
Note that's the flag in the failing test: `Test py3.12 : fast-compile 1 `
This is a bit like the `floatX="float32"`, because it's not the default you may be surprised the test is failing remotely but not locally.

I think this happens because in fast compile we're running the python implementation of `SpecifyShape` and in the default mode we're running the C implementation which have slightly different error messages. Can you make the match statement handle both cases?

**Comment by ricardoV94** [comment]
You seem to be picking commits from main. That probably means you are doing the default pull, which is a pull --merge, instead of `pull --rebase` which would re-apply your commits on top of the main commits, and once you `push --force` won't show those as being commits from this PR

**Review by ricardoV94 — APPROVED** [review]
